### PR TITLE
Adds Additional Support for the Razer Huntsman V2

### PR DESCRIPTION
**Pull request by nightsky30**
Adds Additional Support for the Razer Huntsman V2
Original changes from #1674

Added Functionality:
- [x] Poll Rate (did get packets)
  - It seems to be displaying the same behavior and same values as other devices using `razer_chroma_misc_set_polling_rate2`
  - Up to 8000 Hz
  - Seems Razer added a 250 Hz poll rate.
- [x] Added 250 Hz Poll Rate
- [x] Keyswitch Optimization (did get packets)
  - Typing vs Gaming
  - Implemented as a toggle of state
  - The 1st command (set of two commands sent each time) seems to be toggling the increased debounce delay and faster triggering for the modes.
  - The 2nd command (set of two commands sent each time) is a near copycat that seems to be increasing the number of parameters by 1, using the first parameter as the VARSTORE, and shifting the other parameter indexes by 1.
  - Keyswitch optimization state was verified saved between reboots using D-Feet and Wireshark.

Working (with fix):
- [x] Game Mode LED
  - Seems it works with the Huntsman V2 if the device files are created along with all the other files on the Mouse Protocol interface
  - Tested my changes with both the Huntsman V2 and Blackwidow Chroma V2 which uses one of the additional Keyboard Protocol interfaces
```
qt.qpa.xcb: QXcbConnection: XCB error: 3 (BadWindow), sequence: 1441, resource id: 23079522, major code: 40 (TranslateCoords), minor code: 0
libpng warning: iCCP: known incorrect sRGB profile
Traceback (most recent call last):
  File ’/home/user/Documents/Projects/github/polychromatic/polychromatic/pylib/backends/openrazer.py’, line 1013, in set_device_state
    rdevice.game_mode_led = option_data
  File ’/usr/lib/python3/dist-packages/openrazer/client/devices/keyboard.py’, line 34, in game_mode_led
    self._dbus_interfaces[’game_mode_led’].setGameMode(True)
  File ’/usr/lib/python3/dist-packages/dbus/proxies.py’, line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File ’/usr/lib/python3/dist-packages/dbus/connection.py’, line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Python.PermissionError: Traceback (most recent call last):
  File ’/usr/lib/python3/dist-packages/dbus/service.py’, line 711, in _message_cb
    retval = candidate_method(self, *args, **keywords)
  File ’/usr/lib/python3/dist-packages/openrazer_daemon/dbus_services/dbus_methods/chroma_keyboard.py’, line 85, in set_game_mode
    open(super_file, ’wb’).write(b’\x01’)
PermissionError: [Errno 13] Permission denied: ’/sys/devices/pci0000:00/0000:00:14.0/usb1/1-13/1-13:1.3/0003:1532:026C.0010/key_super’
```

### CODE REVIEW

**Comment by r4v** [comment]
I've compilation error on commit from `nightsky30:feature_huntsman_v2_additional_fixes_functionality` branch 

```
:: Compiling OpenRazer kernel modules
========================================
make -C /lib/modules/5.13.0-27-generic/build M=/media/sdb/tools/openrazer/driver modules
make[1]: Entering directory '/usr/src/linux-headers-5.13.0-27-generic'
  CC [M]  /media/sdb/tools/openrazer/driver/razerkbd_driver.o
  CC [M]  /media/sdb/tools/openrazer/driver/razerchromacommon.o
/media/sdb/tools/openrazer/driver/razerchromacommon.c: In function ‘razer_chroma_misc_set_keyswitch_optimization’:
/media/sdb/tools/openrazer/driver/razerchromacommon.c:947:9: error: label at end of compound statement
  947 |         default:
      |         ^~~~~~~
/media/sdb/tools/openrazer/driver/razerchromacommon.c:963:9: error: label at end of compound statement
  963 |         default:
      |         ^~~~~~~
/media/sdb/tools/openrazer/driver/razerchromacommon.c:967:5: error: label at end of compound statement
  967 |     default:
      |     ^~~~~~~
make[2]: *** [scripts/Makefile.build:281: /media/sdb/tools/openrazer/driver/razerchromacommon.o] Error 1
make[1]: *** [Makefile:1879: /media/sdb/tools/openrazer/driver] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.13.0-27-generic'
make: *** [Makefile:41: driver] Error 2
```

but compiling from `nightsky30:feature_huntsman_v2` build is successful and working flawless

**Comment by nightsky30** [comment]
> I've compilation error on commit from nightsky30:feature_huntsman_v2_additional_fixes_functionality branch

I'm not seeing this on my system, but clearly you and the build checks were having the same issue. I went ahead and removed the default cases. Thanks! :+1:

File: driver/razerkbd_driver.c [hunk 1]
```diff
@@ -2486,6 +2549,79 @@ static ssize_t razer_attr_write_matrix_custom_frame(struct device *dev, struct d
...
+    struct usb_device *usb_dev = interface_to_usbdev(intf);
+    struct razer_report report = razer_chroma_misc_get_polling_rate();
+    struct razer_report response_report = {0};
+    unsigned short polling_rate = 0;
+
+    switch(usb_dev->descriptor.idProduct) {
+    case USB_DEVICE_ID_RAZER_HUNTSMAN_V2:
+        report = razer_chroma_misc_get_polling_rate2();
+        report.transaction_id.id = 0x1f;
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:2569]: add a sensible `default:` or `return -ENOSYS;` or something?

File: driver/razerkbd_driver.c [hunk 2]
```diff
@@ -2486,6 +2549,79 @@ static ssize_t razer_attr_write_matrix_custom_frame(struct device *dev, struct d
...
+    struct usb_interface *intf = to_usb_interface(dev->parent);
+    struct usb_device *usb_dev = interface_to_usbdev(intf);
+    unsigned short polling_rate = (unsigned short)simple_strtoul(buf, NULL, 10);
+    struct razer_report report = razer_chroma_misc_get_polling_rate();
+
+    switch(usb_dev->descriptor.idProduct) {
+    case USB_DEVICE_ID_RAZER_HUNTSMAN_V2:
+        report = razer_chroma_misc_set_polling_rate2(polling_rate);
+        report.transaction_id.id = 0x1f;
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:2617]: add a sensible `default:` or `return -ENOSYS;` or something?

File: driver/razerkbd_driver.c [hunk 3]
```diff
@@ -2486,6 +2549,79 @@ static ssize_t razer_attr_write_matrix_custom_frame(struct device *dev, struct d
...
+ * Write device file "poll_rate"
+ *
+ * Sets the poll rate
+ */
+static ssize_t razer_attr_write_poll_rate(struct device *dev, struct device_attribute *attr, const char *buf, size_t count)
+{
+    struct usb_interface *intf = to_usb_interface(dev->parent);
+    struct usb_device *usb_dev = interface_to_usbdev(intf);
+    unsigned short polling_rate = (unsigned short)simple_strtoul(buf, NULL, 10);
+    struct razer_report report = razer_chroma_misc_get_polling_rate();
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: I assume this should be set and not get?

File: driver/razerkbd_driver.c [hunk 4]
```diff
@@ -552,6 +552,69 @@ static ssize_t razer_attr_read_game_led_state(struct device *dev, struct device_
...
+    // Toggle Keyswitch Optimization
+    switch(usb_dev->descriptor.idProduct) {
+    case USB_DEVICE_ID_RAZER_HUNTSMAN_V2:
+        report = razer_chroma_misc_set_keyswitch_optimization(NOSTORE, mode);
+        report.transaction_id.id = 0x1f;
+        razer_send_payload(usb_dev, &report);
+        report = razer_chroma_misc_set_keyswitch_optimization(VARSTORE, mode);
+        report.transaction_id.id = 0x1f;
+        razer_send_payload(usb_dev, &report);
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:575]: Why set it twice? edit: see comment in razer_chroma_misc_set_keyswitch_optimization function

File: driver/razerchromacommon.c [hunk 5]
```diff
@@ -930,6 +930,78 @@ struct razer_report razer_chroma_misc_fn_key_toggle(unsigned char state)
...
+{
+    struct razer_report report = {0};
+
+    switch(optimization_mode) {
+    case 0x00:
+        // Typing
+        switch(variable_storage) {
+        case NOSTORE:
+            // Typing (Set)
+            report = get_razer_report(0x02, 0x02, 0x04); // calss, id, data size
```
Comment (z3ntu) [inline driver/razerchromacommon.c]: typo calss -> class, also below

File: driver/razerchromacommon.c [hunk 6]
```diff
@@ -930,6 +930,78 @@ struct razer_report razer_chroma_misc_fn_key_toggle(unsigned char state)
...
+/**
+ * Set the keyswitch optimization on the device
+ *
+ * Status Trans Packet Proto DataSize Class CMD Args
+ * 00     1f    0000   00    04       02    02  001400280000               | SET KEY OPTIMIZATION STATE (TYPING SET)
+ * 00     1f    0000   00    05       02    15  010014002800               | SET KEY OPTIMIZATION STATE (TYPING VARSTORE) VARSTORE ????
+ * 00     1f    0000   00    04       02    02  000000000000               | SET KEY OPTIMIZATION STATE (GAMING SET)
+ * 00     1f    0000   00    05       02    15  010000000000               | SET KEY OPTIMIZATION STATE (GAMING VARSTORE) VARSTORE ????
+ */
+struct razer_report razer_chroma_misc_set_keyswitch_optimization(unsigned char variable_storage, unsigned char optimization_mode)
```
Comment (z3ntu) [inline driver/razerchromacommon.c]: This function is quite weird, why are you using VARSTORE as parameter two differentiate between two firmware commands on the device? Why not make a separate function that handles e.g. class 0x02 cmd 0x15?

File: driver/razerkbd_driver.c [hunk 7]
```diff
@@ -552,6 +552,69 @@ static ssize_t razer_attr_read_game_led_state(struct device *dev, struct device_
...
+    struct razer_report response = {0};
+    int state;
+
+    switch(usb_dev->descriptor.idProduct) {
+    case USB_DEVICE_ID_RAZER_HUNTSMAN_V2:
+        report = razer_chroma_misc_get_keyswitch_optimization();
+        report.transaction_id.id = 0x1f;
+        break;
+    default:
+        report = razer_chroma_misc_get_keyswitch_optimization();
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Why not as in other functions set report where the variable is declared to this and remove `default:`?

File: driver/razerkbd_driver.c [hunk 8]
```diff
@@ -2486,6 +2549,79 @@ static ssize_t razer_attr_write_matrix_custom_frame(struct device *dev, struct d
...
+    case 0x08:
+        polling_rate = 1000;
+        break;
+    case  0x10:
+        polling_rate = 500;
+        break;
+    case  0x20:
+        polling_rate = 250;
+        break;
+    case  0x40:
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: remove the double space here and in 0x20 and 0x10

File: daemon/openrazer_daemon/dbus_services/dbus_methods/chroma_keyboard.py [hunk 9]
```diff
@@ -78,11 +78,14 @@ def set_game_mode(self, enable):
...
     driver_path = self.get_driver_path('game_led_state')
 
-    for kb_int in self.additional_interfaces:
-        super_file = os.path.join(kb_int, 'key_super')
-        alt_tab = os.path.join(kb_int, 'key_alt_tab')
-        alt_f4 = os.path.join(kb_int, 'key_alt_f4')
-
+    # Fixes setting Game Mode/LED for Huntsman V2 w/ device
+    # files added to the usual MOUSE PROTOCOL Interface.
+    # Other devices seem to use the KEYBOARD PROTOCOL interface.
```
Comment (z3ntu) [inline daemon/openrazer_daemon/dbus_services/dbus_methods/chroma_keyboard.py]: This comment doesn't make much sense in the file, or reword it to something like "Check the usual MOUSE PROTOCOL interface first, and only then additional interfaces because ..."?

Also are you certain this won't break on other devices?

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 10]
```diff
@@ -487,11 +487,14 @@ class RazerHuntsmanV2(_RippleKeyboard):
     USB_VID = 0x1532
     USB_PID = 0x026C
     HAS_MATRIX = True
+    DEDICATED_MACRO_KEYS = False
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: False is default, no?

**Comment by z3ntu** [comment]
I've already pushed a part of this PR to master: https://github.com/openrazer/openrazer/commit/0ed14ac8262764afbc75a5a3f3c3fb5a0ef8ee6c

File: daemon/openrazer_daemon/hardware/keyboards.py (hunk 10 above)
Comment (nightsky30) [reply daemon/openrazer_daemon/hardware/keyboards.py]: You're right, unnecessary 59eb9ea

File: driver/razerkbd_driver.c (hunk 8 above)
Comment (nightsky30) [reply driver/razerkbd_driver.c]: Oops [ec2e284](https://github.com/openrazer/openrazer/pull/1759/commits/ec2e284e04d9de1a77d8492db804b9a4ef46f58e)

File: driver/razerkbd_driver.c (hunk 3 above)
Comment (nightsky30) [reply driver/razerkbd_driver.c]: Good catch! [2a282ae](https://github.com/openrazer/openrazer/pull/1759/commits/2a282aee5210e09ae18e72439ad98f013e9e3c9a)

File: driver/razerchromacommon.c (hunk 5 above)
Comment (nightsky30) [reply driver/razerchromacommon.c]: Oops [0ce4c4c](https://github.com/openrazer/openrazer/pull/1759/commits/0ce4c4c92fddf884a42c9ed7fce6f80453f6dab6)

File: daemon/openrazer_daemon/dbus_services/dbus_methods/chroma_keyboard.py (hunk 9 above)
Comment (nightsky30) [reply daemon/openrazer_daemon/dbus_services/dbus_methods/chroma_keyboard.py]: I agree, updated to be more clear and less specific to a singular device [69c0db7](https://github.com/openrazer/openrazer/pull/1759/commits/69c0db7c1352e8351ca3208c387fb5cd54463906)

I have not tested every device, but have tested with the Huntsman V2 and the BlackWidow Chroma V2 which uses the keyboard protocol. Both seem to be working fine with this change.

File: driver/razerkbd_driver.c (hunk 7 above)
Comment (nightsky30) [reply driver/razerkbd_driver.c]: That's true, we don't have other keyboards with varying calls at this time. I had used the `default:` statement here as is done in other effect functions. But in those functions they seem to need to call different variations of functions for certain devices. [111120d](https://github.com/openrazer/openrazer/pull/1759/commits/111120d85ee13d593fd21e9adea9a0c5c45780cc)

File: driver/razerkbd_driver.c [hunk 11]
```diff
@@ -565,18 +565,18 @@ static ssize_t razer_attr_write_keyswitch_optimization(struct device *dev, struc
...
     default:
-        report = razer_chroma_misc_set_keyswitch_optimization(NOSTORE, mode);
+        report = razer_chroma_misc_set_keyswitch_optimization_command1(mode);
         report.transaction_id.id = 0x1f;
         razer_send_payload(usb_dev, &report);
-        report = razer_chroma_misc_set_keyswitch_optimization(VARSTORE, mode);
+        report = razer_chroma_misc_set_keyswitch_optimization_command2(mode);
         report.transaction_id.id = 0x1f;
         razer_send_payload(usb_dev, &report);
     }
```
Comment (nightsky30) [inline driver/razerkbd_driver.c]: Looking at this again, I was questioning the need for this switch case block at all at this point. This is the only device making use of this feature at the moment. There are other Huntsman V2 keyboards that state in their manuals they support key optimization. It seems likely they might use the exact same commands. I can remove the switch block if you want.

File: driver/razerchromacommon.c (hunk 6 above)
Comment (nightsky30) [reply driver/razerchromacommon.c]: It is weird. I believe I started out with two functions before the original PR was submitted, but then tried to consolidate them. I've separated the functions out as requested. I can change the names if something else is preferred. [04d300d](https://github.com/openrazer/openrazer/pull/1759/commits/04d300df0592617ee185d8b2282e0242aab24975)

File: driver/razerkbd_driver.c (hunk 4 above)
Comment (nightsky30) [reply driver/razerkbd_driver.c:575]: See https://github.com/openrazer/openrazer/pull/1759#discussion_r933922102

File: driver/razerkbd_driver.c [hunk 12]
```diff
@@ -552,6 +552,67 @@ static ssize_t razer_attr_read_game_led_state(struct device *dev, struct device_
...
+        report.transaction_id.id = 0x1f;
+        razer_send_payload(usb_dev, &report);
+        break;
+    default:
+        report = razer_chroma_misc_set_keyswitch_optimization_command1(mode);
+        report.transaction_id.id = 0x1f;
+        razer_send_payload(usb_dev, &report);
+        report = razer_chroma_misc_set_keyswitch_optimization_command2(mode);
+        report.transaction_id.id = 0x1f;
+        razer_send_payload(usb_dev, &report);
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Right now the default doesn't make any sense because it's the same as huntsman v2 case. Did you mean to not set the transaction id here maybe?

File: driver/razerkbd_driver.c (hunk 11 above)
Comment (z3ntu) [reply driver/razerkbd_driver.c]: generally it's ok, just see other comment I made here

File: driver/razerkbd_driver.c (hunk 12 above)
Comment (nightsky30) [reply driver/razerkbd_driver.c]: I went ahead and removed that default if that's ok. I could also put it back and simply remove the transaction IDs. But there aren't any known devices that use those commands without the transaction ID of ox1f.

Comment (z3ntu) [reply driver/razerkbd_driver.c]: I've added this, I think it's a good idea
```c
    default:
        return -ENOSYS;
```

**Review by z3ntu — APPROVED** [review]
LGTM, thanks for the pull request, as always! :wink:
